### ISSUE TITLE
Pruning a subset of the layers

### ISSUE THREAD

**Issue by phantaurus** [NVIDIA/TensorRT-Model-Optimizer#176]
Hello!

I'm wondering if ModelOpt supports pruning only a subset of layers. Specifically, if it's possible to target only certain types of layers or specify individual layer names. I’ve been reviewing the documents and Jupyter notebook examples but haven’t found any API or configuration options that allow for this kind of selective pruning.

Thank you so much!

**Comment by kevalmorabia97** [comment]
Are you trying to prune a vision model using FastNAS algorithm?

**Comment by kevalmorabia97** [comment]
You can customize rules for search space: https://nvidia.github.io/TensorRT-Model-Optimizer/reference/generated/modelopt.torch.nas.search_space.html#modelopt.torch.nas.search_space.generate_search_space

These updated rules can go with pruning mode: https://nvidia.github.io/TensorRT-Model-Optimizer/guides/2_pruning.html#customizing-pruning-config

You can check the search space to make sure only the layers you want to prune as showing as searchable

**Comment by phantaurus** [comment]
Thank you so much for your reply!

Yes, my model includes both a Conv backbone and a ViT module that follows it. For now, I’m only looking to apply pruning to the Conv backbone.

I’m still relatively new to model pruning, and I’ve noticed there are several tools available, such as Torch-Pruning. Could you help me understand the differences between these tools, particularly in terms of functionality and deployment compatibility? Since I use TensorRT for model deployment, I’m assuming ModelOpt might offer better integration?

**Comment by kevalmorabia97** [comment]
I have not used TorchPruning specifically so cant comment on which is better / easier to use and it also depends on your use cases. For ModelOpt's FastNAS pruning, its a lightweight pruning that uses sensitivity analysis using validation score function to identify less important parameters and then prunes them via symbolic tracing of the pytorch computation graph. It takes Pytorch input and Pytorch output that you can use to deploy with any framework as long as original unpruned model can be deployed on that framework

**Comment by phantaurus** [comment]
Thank you so much for your message! Your introduction was incredibly helpful, and I believe ModelOpt is a great fit for my needs.

I do have another question: I’m planning to explore both pruning and sparsification. Would you recommend a serialized pipeline, applying pruning first, then fine-tuning, followed by sparsification and another round of fine-tuning?

Alternatively, since sparsification can be considered a form of pruning itself, would it make sense to combine structured pruning (e.g., removing entire weight channels) with unstructured sparsification (e.g., pruning 2 out of every 4 elements) and perform a single fine-tuning phase at the end?

**Comment by kevalmorabia97** [comment]
Pruning and Sparsity are orthogonal to each other so I'd recommend just using pruning and see what kind of latency vs accuracy tradeoff you see and later you can consider quantization on top of it for further speedup

**Comment by phantaurus** [comment]
Thank you so much for your insightful suggestions! I'm now experimenting with a combination of pruning and quantization based on your guidance.

I have two follow-up questions about pruning:
The first is that my model doesn’t produce softmax outputs. It’s more of a first-stage model that generates intermediate features for a downstream model. As a result, I evaluate the model based solely on a loss function applied to its output.

Would it be reasonable in this case to use the loss itself as the `score_func`? Since lower loss indicates better performance, but the `score_func `seems to hold a “higher is better” assumption. If so, maybe I need to transform my loss into a "higher is better" format, such as by subtracting it from a large constant?
 
The second question is fine-tuning after pruning. After pruning, would you recommend starting with standard fine-tuning, or going directly into fine-tuning with distillation? Maybe the instinct is to begin with fine-tuning, and only introduce distillation if performance doesn't meet expectations?

**Comment by kevalmorabia97** [comment]
You can use negative loss instead of scoring function.
Generally distillation from unpruned model is always better than fine-tuning without distillation so unless you are limited by resources to use distillation, you should always use distillation

**Comment by phantaurus** [comment]
Thank you so much! I’ve been following your guidance to prune my model.

There is another thing I'd want to confirm: my model consists of two parts — a residual convolutional backbone and an encoder-decoder transformer. The loss is computed after the transformer, but for now, I want to prune just the Conv backbone.

I’ve configured the pruning setup to target only nn.Conv2D layers. However, regarding the constraint setting, if I set prune_constraints = {"flops": "60%"}, does the 60% FLOPs pruning goal apply to the entire model, or only to the layers included in the pruning config?

I guess it only applies to the selected layers, but I’d appreciate your confirmation.

**Comment by kevalmorabia97** [comment]
The pruning constraint apply to the whole model regardless of how much is pruneable. You can profile search space to see what centroid and min subnets look like and based on that decide the constraint (either as a percentage or absolute value)

**Comment by phantaurus** [comment]
I see. Thank you so much!

In my implementation, I inherited nn.Conv2D and created a module called WSTDConv2D. However, modelopt didn't capture this module as well as its parameters. Do I need to register it manually?

**Comment by phantaurus** [comment]
Also I noticed this:

> The model needs to be traceable with ModelOpt’s [torch.fx](https://pytorch.org/docs/stable/fx.html)-like tracer.
> 
> If not, you will see errors or warnings when you run [mtn.convert()](https://nvidia.github.io/TensorRT-Model-Optimizer/reference/generated/modelopt.torch.nas.conversion.html#modelopt.torch.nas.conversion.convert). Note that some of these warnings may not affect the search space and hence can be ignored.
> 
> Note that in some cases certain layers cannot be traced and, if possible, you should adjust their definition and forward method to be traceable. Otherwise, such layers and all affected layers will be ignored in the conversion process.


I did encounter warnings in some of my operations, which can not be converted to torch.fx, such as :
```
for i in range(x.shape[1]): or 
zero_data = torch.zeros(x.shape[1], ...)
```

I'm wondering whether this type of tracing only impacts modelopt's ability to determine the search space, or if it also affects modelopt's ability to generate correct outputs.

From what I’ve observed, the code above doesn’t seem to restrict the search space, such as the input and output dimensions of convolution layers, so I assume modelopt can still generate all valid input/output dimensions for each layer. However, I'm uncertain whether this affects inference correctness, or modelopt's ability to accurately identify less important channels.

Thanks!

**Comment by kevalmorabia97** [comment]
For custom conv, can you add it https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/modelopt/torch/nas/modules/conv.py#L34 and https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/modelopt/torch/trace/modules/nn.py#L128-L130

and see if that shows up in the search space?

**Comment by kevalmorabia97** [comment]
If a module is not torch.fx traceable, that module will not be pruned hence other parameters that go as input / output to this layer will also be not pruned. You can try re-writing forward in some cases where fix is easy (e.g. using fixed number instead of x.shape[1]) and see if that improves the search space

Yes all pruned variants in search space will generate valid outputs

**Comment by kevalmorabia97** [comment]
@phantaurus if your issues are resolved, please close this ticket

**Comment by namgyu-youn** [comment]
> Thank you so much for your reply!
> 
> Yes, my model includes both a Conv backbone and a ViT module that follows it. For now, I’m only looking to apply pruning to the Conv backbone.
> 
> I’m still relatively new to model pruning, and I’ve noticed there are several tools available, such as Torch-Pruning. Could you help me understand the differences between these tools, particularly in terms of functionality and deployment compatibility? Since I use TensorRT for model deployment, I’m assuming ModelOpt might offer better integration?

@phantaurus ; Hi, I hope this comment will be helpful for comparing Torch-Pruning with TensorRT-Model-Optimizer.

First, TensorRT-Model-Optimizer prunes based on Neural Architecture Searching (NAS) methodology. The following figure would be helpful for understanding NAS:

![Image](https://github.com/user-attachments/assets/79b51086-7411-4598-a3eb-f75586cfa92f)

"Search space" means **which arguments are the best choice for the layer** (e.g., in ViT, adjusting patch size or stride could be options). In other words, it **fixes the fundamental layer** (i.e., Conv, Batch Noramlization, Pooling) and finds the best arguments; "Performance estimiation strategy" and "Search strategy" literally means "Performance estimiation" and "Where to move in next step"

Unlike TensorRT-Model-Optimizer, DepGraph (a.k.a., Torch-Pruning) introduces a "**Dependency** Graph" for the dependency. For example, if there is tensor multiplication in the neural network, both tensor's shape is fixed making dependency. Likewise, since batch normalization shares input channels in mini-batches, when you want to prune a BN layer, you should approach channel-level pruning.

If you want to learn more about DepGraph, please check [blog - Code Review](https://until.blog/@namgyu-youn/-codereview--depgraph---towards-any-pruning), or [blog - Paper Review](https://until.blog/@namgyu-youn/-paper--deepgraph---towards-any-structural-pruning-rdre2g47)
